zbus/src/address.rs

```diff
@@ -674,11 +675,19 @@ impl Display for Address {
     }
 }
 
-impl FromStr for Address {
-    type Err = Error;
+/// A parsed bus address
+#[derive(Debug, PartialEq, Eq)]
+pub struct Parsed<'a> {
```

Comment:
`Address` is already supposed to be parsed so this feels like a wrong solution. Any other ideas? I'll think about it too..

Reply:
I thought Address is more "concrete" than Parsed, but I am open to naming suggestions or other proposals :)

Reply:
Hmm? What I wanted here is a way to parse the guid= part, and possibly other options... We would need to break the whole Address API to add fields to every variants. Not my call. This intermdiary type is good enough and flexible, it should fit most cases.

An future alternative may be to completely change the way Address is implemented, to make it opaque and have various accessors. I am not sure this is better either.

Reply:
> We would need to break the whole Address API to add fields to every variants. Not my call.

No we don't. That's what I'm saying. Because enum is non_exhaustive, we can add new variants w/o breaking any API. Hence my suggestion.

> This intermdiary type is good enough and flexible, it should fit most cases.

As I wrote, it's not a good API. It'll only confuse users (who are unfortunate enough to deal with these types directly).

> An future alternative may be to completely change the way Address is implemented, to make it opaque and have various accessors. I am not sure this is better either.

For sure. Please create an issue for that with `api break` label for that.

Reply:
It's essentially a hard API break since it will no longer returns Address:X(_) but Address::XWithGuid()...

Reply:
> It's essentially a hard API break since it will no longer returns Address:X(_) but Address::XWithGuid()...

It's not. non_exhaustive means we can return any variant, even ones we add later. It's basically an escape hatch. In case of `Error` type, we even have to have it. There is no way, we can predict all error variants in advance.

> and it's bad design anyway, what if we have another option?

Sure but still better than introducing a weird concept. `Address` **is** the parsed type. We should keep in mind that most users don't deal with addresses directly.

Reply:
> The more I think about it, the more I think the introduced intermediary "Parsed" type is good design.

You're alone in that. :) However, before we fight each other to death on this, let's first check if guid handling is really that important that it can't wait till the next API break (I'm starting to feel that's not far off at all). The spec says:

> A server may specify a key-value pair with the key guid and the value a hex-encoded 16-byte sequence. [the section called “UUIDs”](https://dbus.freedesktop.org/doc/dbus-specification.html#uuids) describes the format of the guid field. If present, this UUID may be used to distinguish one server address from another.

So this will only affect busd but even then, I'm not convinced it needs to. At least at the moment, it only supports 1 transport per run.

Reply:
If busd can't properly support listening on multiple sockets at the same time until the next zbus API break, I really don't mind at all.

Reply:
> So this will only affect busd but even then, I'm not convinced it needs to. At least at the moment, it only supports 1 transport per run.

No, the server should generate it. The client should parse it and distinguish different connections.

Reply:
> No, the server should generate it. The client should parse it and distinguish different connections.

should, yes. Please address my points above, specifically if this really needs to be addressed before 4.0. See #395 for another reason, we might want an API break soon.

Reply:
Not sure either what to do here. Is this required before 4.0? No. Does this require an API break? No. Can we make it fit in the current Address? Not without breaking every user of Address by having new variants for all existing variants...

You seem to dislike the intermediary address::Parsed, while it offers the required flexibility for extra key=value handling. What to do? If the naming isn't good, let's call it ServerAddress ? we can then rename Address->Transport and alias/deprecate Address. That would fit well with the spec.

Reply:
> Not sure either what to do here.

You answer yourself in the next para. :)

> You seem to dislike the intermediary address::Parsed

Indeed. As I tried to explain, it's totally unnecessary, especially if we can break API. 

> If the naming isn't good

Naming is just a symptom. The real problem is that it's unncessary. It's especially bad because it's exposed to public API. Why does the user need to care? They give you a string and you parse it to an address. Then they use the Address to create a Connection.

Reply:
> I iterated, are you ok with the current proposal? If not, what do you propose?

I was on vacation and now at GUADEC so didn't get a chance to look at your code yet. If you have addressed this, please resolve the thread. I can always re-open if needed.

Reply:
See  [♻️ zv: rename Parsed->ServerAddress and Address->Transport](https://github.com/dbus2/zbus/pull/390/commits/f5b43a9c537a482c6f3d9eacb394145273d37a24)